src/Microsoft.OData.Client/DeepInsertSaveResult.cs

```diff
@@ -0,0 +1,229 @@
+            }
+
+            deepInsertRequestMessage.SetRequestStream(deepInsertRequestMessage.CachedRequestStream);
+
+            try
+            {
+                this.batchResponseMessage = this.RequestInfo.GetSynchronousResponse(deepInsertRequestMessage, false);
+            }
+            catch (DataServiceTransportException ex)
+            {
+                InvalidOperationException exception = WebUtil.GetHttpWebResponse(ex, ref this.batchResponseMessage);
+
+                throw exception;
+            }
+        }
+
+        /// <summary>
+        /// This method processes all the changed descriptors in the entity tracker.
+        /// It loops through all the descriptors and creates relationships between the descriptors if any.
+        /// </summary>
+        /// <typeparam name="T">The type of the top-level resource we want to do a deep insert.</typeparam>
+        /// <param name="descriptors">The list of descriptors in the entity tracker.</param>
+        /// <param name="isRootObject">True if the <paramref name="objects"/> are root or top-level objects, i.e., they were passed directly by the client, otherwise false.</param>
+        /// <param name="objects">Objects of the top-level type we want to perform a deep insert on.</param>
+        /// <exception cref="InvalidOperationException"></exception>
+        internal void BuildDescriptorGraph<T>(IEnumerable<Descriptor> descriptors, bool isRootObject, params T[] objects)
+        {
+            foreach (object obj in objects)
+            {
+                EntityDescriptor topLevelDescriptor = this.RequestInfo.EntityTracker.GetEntityDescriptor(obj);
+
+                // If it's a top-level object sent by the user.
+                if (isRootObject)
+                {
+                    // Validate that we can only have one top level entity in Deep Insert.
+                    if (bulkUpdateGraph.TopLevelDescriptors.Count > 0)
+                    {
+                        throw Error.InvalidOperation(Strings.Context_DeepInsertOneTopLevelEntity);
+                    }
+
+                    bulkUpdateGraph.AddTopLevelDescriptor(topLevelDescriptor);
+                }
+
+                foreach (Descriptor descriptor in descriptors)
+                {
+                    if (descriptor is EntityDescriptor entityDescriptor)
+                    {
+                        if (entityDescriptor.ParentEntity != null &&
+                            entityDescriptor.ParentEntity.Equals(topLevelDescriptor.Entity) &&
+                            !bulkUpdateGraph.Contains(entityDescriptor))
+                        {
+                            // We don't support delete and update in deep insert.
+                            if (entityDescriptor.State == EntityStates.Deleted || entityDescriptor.State == EntityStates.Modified)
+                            {
+                                throw Error.InvalidOperation(Strings.Context_DeepInsertDeletedOrModified);

```

Comment:
I propose including the ID of the of offending entity to the error message to make it easier for the user to debug and track down which entity caused the exception.

Reply:
This is a non-trivial fix. Also  ID alone cannot determine a specific entity since entities might share Ids e.g Customer(1), Order(1)

Reply:
@KenitoInc When I say ID, I meant including the entity type, just like you have shown: `Customer(1)` or `Order(1)`

Reply:
I'm curious about the effort involved in getting the ID here. Since we have the entity descriptor, could we not use the `descriptor.Identity` url property then (if it exists)? Since it's basically an identifier that can be used to narrow down the specific entity that causes the exception.

Reply:
Fixed